docker/manager/_internal/_gen.py

```diff
@@ -0,0 +1,309 @@
+from __future__ import annotations
+
+import os
+import typing as t
+import logging
+import itertools
+from uuid import uuid1
+from typing import TYPE_CHECKING
+from pathlib import Path
+from functools import partial
+from functools import lru_cache
+from collections import defaultdict
+
+import fs
+import yaml
+import cattr
+from yamlinclude import YamlIncludeConstructor
+
+from ._make import CUDACtx
+from ._funcs import send_log
+from ._funcs import render_template
+from .exceptions import ManagerGenerateFailed
+from ._configuration import DOCKERFILE_BUILD_HIERARCHY
+from ._configuration import SUPPORTED_ARCHITECTURE_TYPE
+
+DOCKER_DIRECTORY = Path(os.path.dirname(__file__)).parent.parent
+
+if TYPE_CHECKING:
+    from fs.base import FS
+
+    from .groups import Environment
+
+    IncludeMapping = t.Dict[str, str]
+
+YamlIncludeConstructor.add_to_loader_class(
+    loader_class=yaml.FullLoader,
+    base_dir=DOCKER_DIRECTORY.joinpath("manager", "_internal").__fspath__(),
+)
+
+
+CUDA_ARCHITECTURE_PER_DISTRO = {
+    "debian11": ["amd64", "arm64v8"],
+    "debian10": ["amd64", "arm64v8"],
+    "ubi8": ["amd64", "arm64v8", "ppc64le"],
+}
+
+ARCHITECTURE_PER_DISTRO = {
+    "alpine3.14": SUPPORTED_ARCHITECTURE_TYPE,
+    "debian11": SUPPORTED_ARCHITECTURE_TYPE,
+    "debian10": SUPPORTED_ARCHITECTURE_TYPE,
+    "ubi8": SUPPORTED_ARCHITECTURE_TYPE,
+    "amazonlinux2": ["amd64", "arm64v8"],
+}
+
+
+def walk_include_dir(
+    include_fs: FS,
+    templates_dir: str,
+    checker: t.Iterable[t.Any],
+    filter_key: t.Callable[[str], t.Any] = lambda x: x,
+) -> IncludeMapping:
+    if isinstance(checker, str):
+        checker = [checker]
+
+    results = {}
+
+    templates_fs = include_fs.makedirs(templates_dir, recreate=True)
+    for check, p in itertools.product(
+        checker, templates_fs.walk.files(filter=["*.yaml"])
+    ):
+        if check in p:
+            p = p.strip("/").strip(".yaml")
+            results[filter_key(p)] = f"!include include.d/{templates_dir}/{p}.yaml"
+    return results
+
+
+def unpack_include_content(include_: IncludeMapping, indent: int = 0) -> str:
+    mem_fs = fs.open_fs("mem://")
```

Comment:
Seems like you really want to just use `io.StringIO` or `io.BytesIO` depending on whether pyyaml supports string i/o here.